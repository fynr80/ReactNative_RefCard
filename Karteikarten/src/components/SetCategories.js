/**
 * @file Get and Return Categories & Style for HomeScreen
 * @author Fatih Yanar fatih_yanar@gmx.de
 */

import * as React from 'react';
import {
    View,
    Text,
    TouchableOpacity,
    ImageBackground
} from "react-native";
import {useEffect, useState} from "react";

//import {LinearGradient} from 'expo-linear-gradient';
import firebase from "firebase/compat";
import {useNavigation} from '@react-navigation/native';
import IconF from 'react-native-vector-icons/FontAwesome'

import stylejs from "../style";


const SetCategories = (props) => {
    const id = props.id
    const categoriesName = props.name
    const categoryColor = props.categoryColor;
    const navigation = useNavigation();
    const [themeCounter, setThemeCounter] = useState(0);


    useEffect(() => {
        return getThemeCounter()
    })

    /**
     * Get the amount of Themes of each Category
     */
    const getThemeCounter = () => {
        firebase.firestore()
            .collection('category')
            .doc(props.id)
            .collection("set")
            .onSnapshot(snapshot => {
                const arr = snapshot.docs.map((doc) => ({
                    id: doc.id,
                    name: doc.data().titel
                }))
                setThemeCounter(arr.length)
            })
    }

    /**
     * Got the SearchScreen of the selected Category
     */
    const goToSearch = () => {
        navigation.navigate("Search", {id, categoriesName, categoryColor})
    }

    return (
        <TouchableOpacity
            onPress={goToSearch}
        >
            <View testID="homescreenSet" style={[stylejs.items_Container,{height:120, backgroundColor: categoryColor.left}]}>
            {/* <LinearGradient style={[stylejs.items_Container,{height:120}]} colors={[categoryColor.left,categoryColor.right]}>*/}
                <ImageBackground  source={require('../asserts/intelligence.png')} imageStyle={{width:'30%',position:'absolute',left:100,opacity:0.3}} style={{width: '100%', height: '100%',flexDirection:'row',justifyContent:'space-between'}}>
                    <View >
                        <Text testID={"homescreenSetTitle"} numberOfLines={2} style={stylejs.items_titleText}>{props.name}</Text>
                        <Text style={stylejs.items_amountText}>{themeCounter} Sets</Text>
                    </View>
                    <View>
                        <IconF name='chevron-right' size={35} color='white'/>
                    </View>
                </ImageBackground>
            {/*  </LinearGradient>*/}
            </View>
        </TouchableOpacity>
    )
};

export default SetCategories;
