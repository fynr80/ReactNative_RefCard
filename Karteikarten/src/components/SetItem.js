/**
 * @file SetItem of RefCard app
 * @author Kaan Eray <kaan.eray@mni.thm.de>
 */


import * as React from 'react';
import {
    View,
    Text,
    TouchableOpacity,
    ImageBackground
} from 'react-native';
import {useState, useEffect} from "react";

import {useNavigation} from '@react-navigation/native';
//import {LinearGradient} from 'expo-linear-gradient';
import firebase from "firebase/compat";
import Icon from "react-native-vector-icons/MaterialCommunityIcons";

import stylejs from "../style";

const SetItem = (props) => {
    const item_id = props.item_id;
    const category_id = props.category_id;
    const category_name = props.category_name;
    const categoryColor = props.categoryColor;
    const itemName = props.text;
    const item_description = props.description;
    const [questionCounter, setQuestionCounter] = useState(0);
    const navigation = useNavigation();

    useEffect(() => {
        getQuestionCounter();
    }, [])

    /**
     * Gets the amount of Questions of the given Sets
     */
    const getQuestionCounter = () => {
        firebase.firestore()
            .collection('category')
            .doc(props.category_id)
            .collection("set")
            .doc(props.item_id)
            .collection('card')
            .onSnapshot(snapshot => {
                const arr = snapshot.docs.map((doc) => ({
                    id: doc.id,
                    name: doc.data().titel
                }))
                setQuestionCounter(arr.length);
            })
    }

    return (
        <TouchableOpacity
            onPress={() => {
                navigation.navigate('CardDescription', {
                    category_id,
                    item_id,
                    category_name,
                    itemName,
                    item_description,
                    categoryColor,
                    questionCounter
                });
            }}
        >
            <View testID="SearchscreenQuestion" style={[stylejs.items_Container,{backgroundColor: categoryColor.left }]}>
            {/*<LinearGradient style={[stylejs.items_Container]} colors={[categoryColor.left, categoryColor.right]}>*/}
                <ImageBackground source={require('../asserts/intelligence.png')}
                                 imageStyle={{width: '45%', position: 'absolute', left: 100, opacity: 0.3}}
                                 style={{
                                     width: '100%',
                                     height: '100%',
                                     flexDirection: 'row',
                                     justifyContent: 'space-between',
                                 }}>
                    <View style={{flexDirection: "row", flex: 1}}>
                        <View style={{flex: 1}}>
                            <Text testID="SearchscreenQuestionTitle" style={stylejs.items_titleText}>{props.text}</Text>
                            <Text style={stylejs.items_amountText}>{questionCounter} Questions</Text>
                        </View>
                        <TouchableOpacity testID="SetItemNavToCardQueryBTN" onPress={() => {
                            navigation.navigate('CardQuery', {category_id: category_id, item_id, categoryColor});
                        }}>
                            <Icon name='play-circle-outline' size={50} color='white'/>
                        </TouchableOpacity>

                    </View>
                </ImageBackground>
            {/* </LinearGradient>*/}
            </View>
        </TouchableOpacity>
    )
};

export default SetItem;
