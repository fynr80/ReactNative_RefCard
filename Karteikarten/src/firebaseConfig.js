
import firebase from 'firebase/compat/app';

const firebaseConfig = {
    apiKey: "AIzaSyBksswFLr1YB6en-sRf8mqSFle7bhAzSgU",
    authDomain: "swtp-812df.firebaseapp.com",
    databaseURL: "https://swtp-812df-default-rtdb.firebaseio.com",
    projectId: "swtp-812df",
    storageBucket: "swtp-812df.appspot.com",
    messagingSenderId: "44236558384",
    appId: "1:44236558384:web:2f15f6b5fb190634b71df5",
    measurementId: "G-W33L0G7D18"
};
firebase.initializeApp(firebaseConfig);
export {firebase};
