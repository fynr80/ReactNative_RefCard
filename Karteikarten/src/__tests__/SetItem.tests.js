import React from 'react';
import renderer from 'react-test-renderer';
import {fireEvent, render, screen} from '@testing-library/react-native'
import SetItem from "../components/SetItem";
import firebaseConfig from "../firebaseConfig";
import SetCategories from "../components/SetCategories";


const mockedNavigation = jest.fn();
jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native");
    return {
        ...actualNav,
        useNavigation: () => ({
            navigate: mockedNavigation,
            dispatch: jest.fn(),
        }),
    };
});

const setItem = <SetItem
    text="math"
    description="Set for math"
    category_id="2"
    category_name="Calculus"
    item_id="2.2"
    categoryColor=""/>

describe("Category List elements Component", () => {
    it('renders List element correctly', async () => {
        const tree = renderer.create(setItem).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Calls Navigation to CardDescription with Correct Params', async () => {
        const {getByText} = render(setItem);
        const button = screen.getByText("math");
        fireEvent.press(button);
        expect(mockedNavigation).toBeCalledWith("CardDescription", {
            "category_id": "2",
            "item_id": "2.2",
            "category_name": "Calculus",
            "itemName": "math",
            "item_description": "Set for math",
            "categoryColor": "",
            "questionCounter": 0,
        });
    });
    it('Calls Navigation to CardQuery with Correct Params', async () => {
        const {getByTestId} = render(setItem);
        const button = screen.getByTestId("SetItemNavToCardQueryBTN");
        fireEvent.press(button);
        expect(mockedNavigation).toBeCalledWith("CardQuery", {
            "category_id": "2",
            "item_id": "2.2",
            "categoryColor": "",
        });
    });

    it('Qustion title is correctly', () => {
        const {getByTestId} = render(setItem);
        const title = getByTestId("SearchscreenQuestionTitle");
        expect(title.props.children).toMatch("math");
    });

    it('Question Color is correctly', () => {
        const bgcolor = {
            left: "#2B32B2",
            right: '#1488CC',
        }
        const {getByTestId} = render(<SetItem
            text="math"
            description="Set for math"
            category_id="2"
            category_name="Calculus"
            item_id="2.2"
            categoryColor={bgcolor}
        />);
        const set = getByTestId("SearchscreenQuestion");
        expect(set.props.style[1].backgroundColor).toEqual(bgcolor.left);
    });
})
