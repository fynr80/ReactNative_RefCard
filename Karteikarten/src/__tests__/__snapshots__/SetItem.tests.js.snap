// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category List elements Component renders List element correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 30,
          "flexDirection": "row",
          "height": 180,
          "justifyContent": "space-between",
          "marginTop": 20,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 20,
          "width": "100%",
        },
        Object {
          "backgroundColor": undefined,
        },
      ]
    }
    testID="SearchscreenQuestion"
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Object {
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <Image
        source={1}
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": "100%",
              "width": "100%",
            },
            Object {
              "left": 100,
              "opacity": 0.3,
              "position": "absolute",
              "width": "45%",
            },
          ]
        }
      />
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 20,
                "fontWeight": "bold",
                "maxWidth": 300,
                "paddingRight": 30,
              }
            }
            testID="SearchscreenQuestionTitle"
          >
            math
          </Text>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 15,
                "opacity": 0.8,
              }
            }
          >
            0
             Questions
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="SetItemNavToCardQueryBTN"
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
</View>
`;
